src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+                        else
+                        {
+                            edmDeletedObject.Id = string.Empty;
+                        }
+
+                        CopyInstanceAnnotations(changedObj, edmDeletedObject);
+                        edmDeletedObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);                       
+                        edmChangedObject = edmDeletedObject;
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                case DataModificationOperationKind.Unlink:
+                    {
+                        Type type = typeof(EdmDeltaEntityObject<>).MakeGenericType(typeof(TStructuralType));
+                        EdmDeltaEntityObject edmDeltaEntityObject = Activator.CreateInstance(type, EntityType) as EdmDeltaEntityObject;
+                        foreach(PropertyInfo property in typeof(TStructuralType).GetProperties())
+                        {
+                            edmDeltaEntityObject.TrySetPropertyValue(property.Name, property.GetValue(originalObj));
+                        }
+
+                        CopyInstanceAnnotations(changedObj, edmDeltaEntityObject);
+                        edmDeltaEntityObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
+                        edmChangedObject = edmDeltaEntityObject;
+                        break;
+                    }
+            }
+        }
+
+        private static void CopyInstanceAnnotations(dynamic changedObj, IEntityObjectInstanceAnnotations destinationObject)
+        {
+            //This is for copying both Persistent and Transient Instance Annotations.
+            IEntityObjectInstanceAnnotations entityObject = changedObj as IEntityObjectInstanceAnnotations;
+
+            if (entityObject != null)
+            {
+                destinationObject.PersistentInstanceAnnotationsContainer = entityObject.PersistentInstanceAnnotationsContainer;
+                destinationObject.TransientInstanceAnnotationContainer = entityObject.TransientInstanceAnnotationContainer; 
+            }           
+        }
+
+        private static TStructuralType GetFilteredItem(Type type, IEnumerable<IEdmStructuralProperty> keys, IEnumerable<TStructuralType> originalList, IEdmChangedObject changedObject)
+        {   
+            //This logic is for filtering the object based on the set of keys,
+            //There will only be very few key elements usually, mostly 1, so performance wont be impacted.
+            foreach(IEdmStructuralProperty key in keys)
+            {
+                object obj;
+                if (changedObject.TryGetPropertyValue(key.Name, out obj))
+                {
+                    originalList = originalList.Where(x => type.GetProperty(key.Name).GetValue(x).ToString() == obj.ToString());
```

Comment:
>type.GetProperty(key.Name) [](start = 59, length = 26)

Getting the property accessor each time through the loop is expensive -- we should get it once and re-use